src/samples/audio/detect_test.c

```diff
@@ -231,14 +231,22 @@ static void default_detect_test(struct comp_dev *dev,
 
 	/* perform detection within current period */
 	for (sample = 0; sample < count && !cd->detected; ++sample) {
-		src = (valid_bits == 16U) ?
-		      audio_stream_read_frag_s16(source, sample) :
-		      audio_stream_read_frag_s32(source, sample);
-		if (valid_bits > 16U) {
+		switch (valid_bits) {
+		case 16:
+			src = audio_stream_read_frag_s16(source, sample);
+			diff = abs(*(int16_t *)src) - abs((int16_t)cd->activation);
+			break;
+		case 24:
+			src = audio_stream_read_frag_s32(source, sample);
+			diff = abs(sign_extend_s24(*(int32_t *)src)) - abs(cd->activation);
```

Comment:
Took me again while to doublecheck this module only handles the LSB variant of 24in32, in which case above code is correct.